### PR TITLE
Fix the compatible issue of the Dynamo and the PyDev.Debugger.

### PR DESCRIPTION
**Pull request by chengjunlu**
The PyDev.Debugger use the _PyFrameEvalFunction to debug the python script.
Fallback to the previous _PyFrameEvalFunction to fix the dynamo with PyDev.Debugger issue.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by chengjunlu** [comment]
There is another SIGFAULT issue in the PyDev with this fix.
And it should be fixed in newer PyDev with this PR.
https://github.com/fabioz/PyDev.Debugger/pull/186/files

A temp work around in PyTorch is that not to save the code object cache in the co_extra of the PyCodeObject.
I am not sure the work around is also preferred in this PR.

File: torch/csrc/dynamo/eval_frame.c [hunk 1]
```diff
@@ -357,7 +360,10 @@ inline static PyObject* eval_frame_default(
   if (tstate == NULL) {
     tstate = PyThreadState_GET();
   }
-  return _PyEval_EvalFrameDefault(tstate, frame, throw_flag);
+  if (previous_eval_frame)
+    return previous_eval_frame(tstate, frame, throw_flag);
+  else
+    return _PyEval_EvalFrameDefault(tstate, frame, throw_flag);
```
Comment (ezyang) [inline torch/csrc/dynamo/eval_frame.c:366]: nit: use braces around the if-else branches

File: torch/csrc/dynamo/eval_frame.c [hunk 2]
```diff
@@ -381,9 +388,9 @@ inline static void enable_eval_frame_shim(PyThreadState* tstate) {
 inline static void enable_eval_frame_default(PyThreadState* tstate) {
 #if PY_VERSION_HEX >= 0x03090000
   if (_PyInterpreterState_GetEvalFrameFunc(tstate->interp) !=
-      &_PyEval_EvalFrameDefault) {
-    _PyInterpreterState_SetEvalFrameFunc(
-        tstate->interp, &_PyEval_EvalFrameDefault);
+      previous_eval_frame) {
+    _PyInterpreterState_SetEvalFrameFunc(tstate->interp,
+                                         previous_eval_frame);
```
Comment (ezyang) [inline torch/csrc/dynamo/eval_frame.c:393]: Should we NULL out `previous_eval_frame` at this point?

**Comment by ezyang** [comment]
This seems broadly OK, just some nits. Also want to give others a chance to look.

It would be nice to have a test for this, though I'm not sure how we can easily test this in CI without actually installing PyDev.Debugger.

`co_extra` clobbering is troublesome. I don't know if there's a way to deal with this in general, unfortunately.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**